src/test/java/com/arpnetworking/metrics/mad/parsers/JsonToRecordParserV2gTest.java

```diff
@@ -0,0 +1,439 @@
+        parseRecord("QueryLogParserV2gTest/testBadGaugeValuesValueDenominatorUnitsName.json");
+    }
+
+    @Test(expected = ParsingException.class)
+    public void testBadTimerValuesValueNumeratorUnitsName() throws ParsingException, IOException {
+        parseRecord("QueryLogParserV2gTest/testBadTimerValuesValueNumeratorUnitsName.json");
+    }
+
+    @Test(expected = ParsingException.class)
+    public void testBadTimerValuesValueDenominatorUnitsName() throws ParsingException, IOException {
+        parseRecord("QueryLogParserV2gTest/testBadTimerValuesValueDenominatorUnitsName.json");
+    }
+
+    @Test
+    public void testNaNValues() throws ParsingException, IOException {
+        final Record record = parseRecord("QueryLogParserV2gTest/testNaNValues.json");
+        Assert.assertNotNull(record);
+
+        Assert.assertEquals(DateTime.parse("2014-03-24T12:15:41.010Z"), record.getTime());
+        Assert.assertEquals(3, record.getDimensions().size());
+        Assert.assertEquals("MyHost", record.getDimensions().get(Key.HOST_DIMENSION_KEY));
+        Assert.assertEquals("MyService", record.getDimensions().get(Key.SERVICE_DIMENSION_KEY));
+        Assert.assertEquals("MyCluster", record.getDimensions().get(Key.CLUSTER_DIMENSION_KEY));
+
+        final Map<String, ? extends Metric> variables = record.getMetrics();
+        Assert.assertEquals(3, variables.size());
+
+        Assert.assertThat(variables, Matchers.hasKey("t1"));
+        final Metric t1 = variables.get("t1");
+        Assert.assertTrue(t1.getValues().isEmpty());
+
+        Assert.assertThat(variables, Matchers.hasKey("g1"));
+        final Metric g1 = variables.get("g1");
+        Assert.assertTrue(g1.getValues().isEmpty());
+
+        Assert.assertThat(variables, Matchers.hasKey("c1"));
+        final Metric c1 = variables.get("c1");
+        Assert.assertTrue(c1.getValues().isEmpty());
+    }
+
+    private static void assertValue(final Quantity quantity, final double value) {
+        assertValue(quantity, value, Optional.empty());
+    }
+
+    private static void assertValue(final Quantity quantity, final double value, final Unit unit) {
+        assertValue(quantity, value, Optional.of(unit));
+    }
+
+    private static void assertValue(final Quantity quantity, final double value, final Optional<Unit> unit) {
+        Assert.assertEquals(value, quantity.getValue(), 0.001);
+        if (unit.isPresent()) {
+            Assert.assertTrue(quantity.getUnit().isPresent());
+            Assert.assertEquals(unit.get(), quantity.getUnit().get());
+        } else {
+            Assert.assertFalse(quantity.getUnit().isPresent());
+        }
+    }
+
+    private static Record parseRecord(final String fileName) throws ParsingException, IOException {
+        return new JsonToRecordParser.Builder()
+                .build()
+                .parse(Resources.toByteArray(Resources.getResource(
+                        JsonToRecordParserV2gTest.class, fileName)));
+    }
+}
```

Comment:
The tests for versions 2c, 2d, 2e and 2f need updating as they need to validate that they produce the map to the correct record definition in all those use cases.